Add stop support to openzwave (mqtt) cover

clook opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

Not sure how how it may have a negative impact / breaking change over the current implementation.


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add stop support to openzwave (mqtt) cover
Attempts to solve https://github.com/home-assistant/core/issues/38018


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #38018
- This PR is related to issue: 
- Link to documentation pull request: 

Button press and release are not implemented in python openzwave mqtt, so boolean constants have been used instead. Not sure about the location of these constants as button press and release implementation should be the business of openzwave mqtt library anyway.

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

clook committed:
feat: add stop to openzwave (mqtt) cover

homeassistant commented:
Hi @clook, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#44622).

Once you do that we will be able to review and accept this pull request.

Thanks!

clook committed:
Fix isort and black linter

homeassistant commented:
Hi @clook, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#44622).

Once you do that we will be able to review and accept this pull request.

Thanks!

MartinHjelmare commented on homeassistant/components/ozw/cover.py:
```diff
@@ -52,6 +57,12 @@ def percent_to_zwave_position(value):
+        super().__init__(values)
+
     @property
     def supported_features(self):
```
We can remove this, since the default cover entity support now matches the support we need.

clook replied:
Nice catch! Removed everything related.

MartinHjelmare commented on homeassistant/components/ozw/cover.py:
```diff
@@ -73,11 +84,22 @@ async def async_set_cover_position(self, **kwargs):
     async def async_open_cover(self, **kwargs):
         """Open the cover."""
-        self.values.primary.send_value(99)
+        self.values.open.send_value(PRESS_BUTTON)
```
How does this impact covers that used the primary value before? Do all devices support the button presses?

clook replied:
When checking to `zwave/cover.py` component, the button press is the behavior used. Sending a extremum value was a workaround while button values were not supported through openzwave / mqtt.
What could be wrong in the current implementation is the lack of support for inverted buttons or cover position (it is supported by `zwave/cover.py`), if we find a device where buttons are inverted but the position is not.

MartinHjelmare replied:
Someone should verify that the behavior is the same between OpenZWave 1.4 and 1.6.

kpine replied:
> How does this impact covers that used the primary value before? Do all devices support the button presses?

All window coverings (not garage doors) are required to support the Multilevel Switch Command Class. The `open` and `close` values are [mapped by HA](https://github.com/home-assistant/core/blob/3de0610909e97f92a09a9efae117e0fabfeb0067/homeassistant/components/ozw/discovery.py#L183) (for both ozw and zwave) to the button values that OZW creates called "Bright" and "Dim", which then correspond to the  Z-Wave Multilevel Switch commands Start Level Change and Stop Level Change. Therefore all window coverings support this button API, since the Start/Stop Level Change commands have existed since V1. 

The primary value would be the "Level" value itself, which sets the "position" (brightness).

> Someone should verify that the behavior is the same between OpenZWave 1.4 and 1.6.

The implementations of the button values are nearly identical ([1.4 hass](https://github.com/home-assistant/open-zwave/blob/hass/cpp/src/command_classes/SwitchMultilevel.cpp#L290) vs [1.6](https://github.com/OpenZWave/open-zwave/blob/4fe9af833a6d1fc8949918bf117e131378b7d059/cpp/src/command_classes/SwitchMultilevel.cpp#L250)). 

A button press with the `open` value triggers a Start Level Command with the direction "up". A button press with the `close` value triggers a Start Level Command with the direction "down". A button release will issue a Stop Level Command. It doesn't matter which button you use for the "release" action, either one triggers the same result, which is why the zwave integration seems to have picked the `open` value for stop (or close value if inverted).

The implementation of `SwitchMultilevel::StartLevelChange` (which takes the up or down direction) looks about the same to me.

As for ozwdaemon, it [uses](https://github.com/OpenZWave/qt-openzwave/blob/master/docs/MQTT.md#setvalue) true for a button press and false for a release, and [calls](https://github.com/OpenZWave/qt-openzwave/blob/7ebd43e246e97851ad20f2b1bb9c62bfd6a4a6ef/qt-openzwave/source/qtozwmanager_p.cpp#L2206) the same pressButton and releaseButton API that the zwave integration uses.

MartinHjelmare replied:
@kpine thank you! Great insight!

clook committed:
Remove supported_features for cover.

As suggested by @MartinHjelmare, not needed anymore because base class
implementation is sufficient.

https://github.com/home-assistant/core/pull/44622#discussion_r549854542

clook commented on homeassistant/components/ozw/cover.py:
```diff
@@ -73,11 +74,22 @@ async def async_set_cover_position(self, **kwargs):
+        self.values.close.send_value(PRESS_BUTTON)
+        self._close_button_pressed = True
+
+    async def async_stop_cover(self, **kwargs):
+        """Stop cover."""
+        if self._open_button_pressed:
+            self.values.open.send_value(RELEASE_BUTTON)
+            self._open_button_pressed = False
+        if self._close_button_pressed:
+            self.values.close.send_value(RELEASE_BUTTON)
+            self._close_button_pressed = False
```
There seems to be a kind of idempotency on button press / release on openzwave / mqtt, which makes the "close" press not working if we only release the "open", hence the additional tracking of the pressed status of a button (a bit hacky). On the legacy `zwave/cover.py` implementation, using a release on the "open" button was sufficient.

What we can try to make the code simpler, is to send release on both buttons regardless of any tracking. It should make it work (and idempotency should not have any wrong effect then). I will also try scenario with local (physical button) push to see if a remote stop work.

MartinHjelmare commented:
I think this looks good. @clook let me know when you're happy with the testing and we can go ahead here.

clook commented:
Hi, thanks for your feedback @MartinHjelmare and @kpine 

Regarding https://github.com/home-assistant/core/pull/44622#discussion_r549908015, and some testing, as expected:
- the button state (pressed or released) is remembered by lower level (Openzwave?) for idempotency, if we issue a release on an already released button, or a push on an already pushed one, it will have no effect
- if we rely on this lower level state management, we could remove state from the PR (`self._open_button_pressed` and `self._close_button_pressed`) and only issue the release of both buttons (we ensure a pushed button gets released, and no useless event will happen if the button is considered as already released)
- for the same reason, only issuing an release open button event here is not sufficient if the pressed button was close
- also, the button release will not work when button has been triggered physically if lower level state is already known as "released". We can trigger the stop event by cheating with a button push event followed by a physical stop and a physical button (so the internal state is still "pushed"). I am not sure if we can change this behavior of the lower level (it is not a big deal to stop a cover that was opened or closed by a physical button push but still a limitation)

I don't like the fact to rely on an upstream behavior that is not clearly documented or an expected contract. We can comment out the code from the PR to indicate we are relying on this behavior though, since it will make it simpler. I will also try to catch where this button state is coded at a lower level if need to point it out.

What do you think about it?

MartinHjelmare commented:
I think it's ok to keep the internal state for the buttons. Add a comment why we use it. Or if you prefer to remove the internal state and comment about the behavior that could be ok too.

Are there any limitations if try to support stopping the cover after a physical device has released the button? It's ok to wait with this too to a potential follow up PR.

clook committed:
Make a simpler version depending on idempotency

qt-openzwave already implements idempotency, see:
https://github.com/OpenZWave/qt-openzwave/blob/77e414217f83fae89a8f41156db3783d562703b1/qt-openzwave/source/qtozwvalueidmodel.cpp#L180

We can use it and trigger button release anywhen.

MartinHjelmare commented on homeassistant/components/ozw/cover.py:
```diff
@@ -73,11 +68,19 @@ async def async_set_cover_position(self, **kwargs):
+        """Stop cover."""
+        # Need to issue both buttons release since qt-openzwave implements idempotency
+        # keeping internal state of model to trigger actual updates. We could also keep
+        # another state in Hass.io to know which button to release, but this implementation is simpler.
```
```suggestion
        # another state in Home Assistant to know which button to release,
        # but this implementation is simpler.
```

MartinHjelmare commented on tests/components/ozw/test_cover.py:
```diff
@@ -16,53 +16,98 @@ async def test_cover(hass, cover_data, sent_messages, cover_msg):
+    assert msg["topic"] == "OpenZWave/1/command/setvalue/"
+    assert msg["payload"] == {"Value": False, "ValueIDKey": 562950578995224}
+
 
```
```suggestion
```

MartinHjelmare committed:
Clean up

MartinHjelmare reviewed (approved):
Thanks!

clook commented:
I finally switch to the simpler implementation, relying on qt-openzwave behaviour. It works also great this way and we don't have 2 distinct tracks of buttons state.
If another implementation of MQTT Zwave is made without the state thing, it shouldn't heart to have a button release event if it is not pressed.

Thanks again for your review @MartinHjelmare and all your great support.
